src/main/java/redis/clients/jedis/JedisClusterInfoCache.java

```diff
@@ -86,21 +86,24 @@ public void discoverClusterSlots(Jedis jedis) {
       rediscovering = true;
 
       try {
-        this.slots.clear();
-
         List<Object> slots = jedis.clusterSlots();
 
+        // We should clear slots after getting result about cluster slots, because if we got exception or timeout in
```

Comment:
>> And IMHO, clearing slots first or later is up to amount of slot changeset

> I'm not sure if I got you. We clear it completely anyways. I just moved clearing after external code. So, to avoid situation of exception in external code, which leads to publishing empty slots collection.

Imagine first trial of clusterSlots() failed (so lock is released), and next one calls getConnectionFromSlot(). If we don't clear slots, it will get Jedis instance from cache which could be invalidated (cause we fails to update so doesn't catch up latest information) or not (node for slot is not changed). Former will try to request wrong node, and receive MOVED, which also initiate slot cache update. If we clear cache first, it doesn't need to request but just update slot cache immediately.